### ISSUE TITLE
Reimplement iForest building in Rust

### ISSUE THREAD

**Issue by hombit** [snad-space/coniferest#356]
We would highly benefit from decoupling our code from scikit-learn and making iForest builds parallel. @matwey and I propose this plan

## Parallel forest building

- Shared immutable training data matrix.
- Each subsample + iTree build forms a parallel job. Each job has its own random seed and borrows the shared data matrix.
- The main thread samples random seeds for each tree build in advance.

## iTree build job

- Each job has a mutable array for storing subsample indices (or references).
  - This may be optimized with thread-local storage or a better allocator.
- This subsample index array is partitioned at each iTree node split.
- The split logic should be customizable.
  - The default implementation would use the original Isolation Forest logic: sample random features, then sample the split value from a uniform distribution between the minimum and maximum values of the selected feature in the current sample.
- Each split task is added to a task queue to avoid deep recursion.
  - `Task { new_node_index: u32, depth: u16, slice: &mut subsample_index_array }`
  - Algorithm:
    1. Create the root task: `Task { new_node_index: 0u32, depth: 0u16, slice: &mut full_subsample_index_array }`, and add it to the task queue.
    2. While the task queue is not empty:
    3. Pop the first task. If `depth < max_depth` and `slice.len() > 1`, call the splitter. Otherwise, create a `Leaf` at `nodes[new_node_index]`.
    4. The splitter takes `slice` and `depth` and returns the split feature, split value, and whether the node should be split or become a leaf, e.g. `Option<(u16, T)>`.
    5. If the splitter indicates that the node should not be split, create a `Leaf` at `nodes[new_node_index]`. Otherwise, create a `SplitNode` at `nodes[new_node_index]` and enqueue two new tasks.
    6. Partition `task.slice` into `left_subsample_index_array` and `right_subsample_index_array` according to the split feature and split value.
    7. Enqueue:
       - `Task { new_node_index: nodes.len(), depth: depth + 1, slice: &mut left_subsample_index_array }`
       - `Task { new_node_index: nodes.len() + 1, depth: depth + 1, slice: &mut right_subsample_index_array }`

## iForest storage

- Each tree is a separate object. The forest is stored as `Vec<Box<Tree>>`, `VecDeque<Tree>`, or a Python list.
  - We should evaluate whether a Python list is suitable for parallel traversal.
- Trees are built in temporary storage, such as `Vec::with_capacity(n_node_max)` or `VecDeque`. Once construction is complete, they are copied into the permanent `ndarray::Array` or `Vec` storage.
- `struct SplitNode { left_node_index: NonZeroU32, split_feature: u32, split_value: T /* f32/f64 */ }`
  - `right_node_index = left_node_index + 1` and may be omitted.
- `struct Leaf { leaf_index: u32, value: f32 /* or T */ }`
  - The leaf value is computed by a customizable algorithm.
- `enum Node { SplitNode(SplitNode), Leaf(Leaf) }`
- `struct Tree { node_vec: Vec<Node> /* or ndarray::Array<Node> */, n_leaves: u32, n_subsamples: usize }`
